midend/flattenInterfaceStructs.h

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2018 VMware, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _MIDEND_FLATTENINTERFACESTRUCT_H_
+#define _MIDEND_FLATTENINTERFACESTRUCT_H_
+
+#include "ir/ir.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+
+namespace P4 {
+
+/**
+   Describes how a type is replaced: the new type to
+   replace it and how each field is renamed.
+*/
+struct Replacement : public IHasDbPrint {
+    // The keys are strings of the form field.field.field.  If
+    // something is mapped to the null string it does not really have
+    // a replacement: this is true for fields whose types are structs.
+    std::map<cstring, cstring> fieldNameRemap;
+    const IR::Type* type;
+    virtual void dbprint(std::ostream& out) const {
+        out << type;
+    }
+};
+
+/**
+ * Maintains a map between an original struct type and its
+ * replacement.
+ */
+struct NestedStructMap {
+    P4::ReferenceMap* refMap;
+    P4::TypeMap* typeMap;
+
+    ordered_map<const IR::Type*, Replacement*> replacement;
```

Comment:
Are you keeping the replacement per field or per struct? From your comment on replacement, I assumed per field.

Reply:
For each type that is to be replaced (a deeply nested struct) we create a Replacement. The replacement says how each field is renamed, and also it has a pointer to the new flattened struct. When I rename Replacement I will give that field a clearer name, now it's called type.